### DISCUSSION TITLE
improve search speed, now it is very slow in CPU, 33 ms per entity.

### DISCUSSION THREAD

**Discussion by rocke-dong** [milvus-io/milvus#16508] [Q&A and General discussion] [answered]
## Environment
- Milvus version: v2.0.2
- Deployment mode(standalone or cluster): standalone 
- SDK version: pymilvus v2.0.2, python 3.8
- OS(Ubuntu or CentOS): Ubuntu 20.04
- CPU/Memory: 8  Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz / 64G

## Problem: search speed is very slow, 33 ms per entity.
milvus configs:
nlist: 600
nprobe: 5
metric_type: "L2"
index_type: "IVF_FLAT"
index_file_size: 2048 (MB)
queryNode:
  cacheSize: 32 # GB, default 32 GB, `cacheSize` is the memory used for caching data for faster query.
data infor:
num entity: 167707
vector dim size: 12038
vector type: float
search entity num: 4209
search total time: 139 seconds
search time per entity num: 139/4209*1000=33 ms!

## My trial
1. I search many articles, and so use very small nprobe, 5.
2. nlist is based on general terms, the recommended value of nlist is 4 × sqrt(n), where n is the total number of entities in a segment.
3. I test index_file_size 1024 and 2048 when initialize Connection, 1024 and 2048 too tiny difference on search times.
```
from pymilvus import Collection
collection = Collection(name=collection_name, schema=schema, using='default', shards_num=2, 
        index_file_size=2048)
```

5. I test batch search batch size = 500 and batch size = 1
that's the data array first dimension size, that's I input 1 entity or 500 entities into collection.search() once
```
collection.search(data=data, anns_field=self.vector_field_name, 
            param=self.mivlus_search_params, limit=self.top_num)
```
batch size 1 use about 1.5 times of time than batch size 500.
So, the batch size has small influence on search speed, the search is not multi process?

## Questions
Could you give me advice to improve the search speed?
If collection.search is not multi process or multi thread, do we need write extra multi process or multi thread?
Another possible solution, use GPU during search, yes?

thanks!!!

**Comment by xiaofan-luan** [comment]
what is your target qps? few things you may want to try:
1. try HNSW instead of Faiss index
2. try larger nq than exact 1 query
3. wait for milvus 2.1(Great performance improvement)

**Reply by rocke-dong** [reply]
thanks!!
my target is total query seconds for one batch items, so 10ms for each item, <10ms better. I think milvus2.* GPU will be much better for my target.
1. I will try HNSW.
2. I test nq 1, 200, 500, 1000. 500 has the best perfermance.
3. I wil wait for milvus2.1.

**Reply by rocke-dong** [reply]
I tested IVF_SQ8, about 2~3 times of IVF_FLAT, much slower speed, but the same recal vs IVF_FLAT.
I tested HNSW, yes the speed is faster,  but recall reduce relatively. I test three group parameter of HNSW.
In a word, HNSW maybe not my choice, because of low recall vs IVF_FLAT.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/qc_do/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/qc_do/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



index_file_size     (MB) | batch size | ef | M | efConstruction | search time     (seconds) | search time per ent     (mili seconds) | full correct num | performance     ratio vs IVF_FLAT
-- | -- | -- | -- | -- | -- | -- | -- | --
2048 | 500 | 4 | 16 | 64 | 21 | 5.0 | 2847 | 0.862727273
  |   | 64 | 16 |   | 72 | 17.1 | 3113 | 0.943333333
  |   | 64 | 64 | 64 | 81 | 19.2 | 3107 | 0.941515152



</body>

</html>

**Reply by xiaofan-luan** [reply]
From our test result, HNSW is far accurate and fast than IVF, I'm not sure if it's because the vector dim is huge because under most case we only test for 768 dim vectors.

And it doesn't make sense to me that IVF_SQ8 is slower. because SQ8 take only 1/4 of the calculation if you are using float vector so usually should be 2-3 times faster than IVF_FLAT.

could you also list the IVF_FLAT search time and accuaracy so we can take a deep look?

**Reply by xiaofan-luan** [reply]
Since you have very small entity numbers, I do believe GPU can help you on the cases.
Please contact me on Slack or wechat group if you want to further discussion

**Reply by rocke-dong** [reply]
> From our test result, HNSW is far accurate and fast than IVF, I'm not sure if it's because the vector dim is huge because under most case we only test for 768 dim vectors.
> 
> And it doesn't make sense to me that IVF_SQ8 is slower. because SQ8 take only 1/4 of the calculation if you are using float vector so usually should be 2-3 times faster than IVF_FLAT.
> 
> could you also list the IVF_FLAT search time and accuaracy so we can take a deep look?

It is also a surprise  IVF_SQ8 is much slower in this case.
IVF_FLAT search time: 140 seconds for 4209 entities, 33ms search time per entity.   Accuaracy, full correct num 3300 items.

**Reply by xiaofan-luan** [reply]
> emm, doesn't make much sense to me. Did you check the index building is working?

**Reply by rocke-dong** [reply]
> > emm, doesn't make much sense to me. Did you check the index building is working?

yes, I surely create_index after inserting data. There is no error for create_index().
could you tell me how to check  the index building is working?
thanks!

**Reply by rocke-dong** [reply]
> > emm, doesn't make much sense to me. Did you check the index building is working?
Could you tell me how to check the index building is working? Is there a special way to check this? I cannot find it in milvus website
thanks!

**Comment by rocke-dong** [comment, marked as answer]
Dear all, 
Mivlus 2.0.2, In sdk, the "index_file_size" parameter doesn't work, when I modify the segment maxSize in milvus.yml from 512 to 2048, the search time reduces from 130 seconds to 41 seconds for 4209 entities, 9.7ms for each entity, number of query is 500,  IVF_FLAT. A big improvement in query speed.

segment:
    maxSize: 512 # Maximum size of a segment in MB

**Reply by yhmo** [reply]
If there is only one query node, big segment is more efficient than small segment.

But for multiple query nodes, we need to carefully specify a proper segment size according to query nodes number. For example, raw data size is 2GB, there are 4 query nodes in milvus cluster, the proper segment size is 2GB/4 = 512MB. Then raw data is split into 4 segments, so that milvus can balance segments to 4 query nodes to get best performance.
